backend/tilavarauspalvelu/models/payment_order/actions.py

```diff
@@ -109,3 +132,73 @@ def update_order_status(self, new_status: OrderStatus, payment_id: str = "") ->
 
         EmailService.send_reservation_confirmed_email(reservation=reservation)
         EmailService.send_reservation_confirmed_staff_notification_email(reservation=reservation)
+
+    def cancel_together_with_verkkokauppa(self, *, cancel_on_error: bool = False) -> None:
+        if self.payment_order.status not in OrderStatus.can_be_cancelled_statuses:
+            return
+
+        if not settings.MOCK_VERKKOKAUPPA_API_ENABLED and self.payment_order.remote_id is not None:
+            try:
+                webshop_order = VerkkokauppaAPIClient.cancel_order(
+                    order_uuid=self.payment_order.remote_id,
+                    user_uuid=self.payment_order.reservation_user_uuid,
+                )
+
+                # Don't update if order is missing or something other than cancelled
+                if webshop_order is None or webshop_order.status != WebShopOrderStatus.CANCELLED:
+                    return
+
+            # If there is an unexpected response or error from verkkokauppa,
+            # payment order should still be cancelled
+            except CancelOrderError as error:
+                SentryLogger.log_message("Verkkokauppa: Failed to cancel order", details=str(error))
+
+                if not cancel_on_error:
+                    raise
+
+        self.payment_order.status = OrderStatus.CANCELLED
+        self.payment_order.processed_at = local_datetime()
+        self.payment_order.save(update_fields=["status", "processed_at"])
+
+    def refund_together_with_verkkokauppa(self) -> None:
+        if self.payment_order.status not in OrderStatus.can_be_refunded_statuses:
+            return
+
+        # Refunds can only be made for orders paid through webshop.
+        if self.payment_order.remote_id is None:
+            return
+
+        if settings.MOCK_VERKKOKAUPPA_API_ENABLED:
+            self.payment_order.refund_id = uuid.uuid4()
+        else:
+            refund = VerkkokauppaAPIClient.refund_order(order_uuid=self.payment_order.remote_id)
+            self.payment_order.refund_id = refund.refund_id
+
+        self.payment_order.status = OrderStatus.REFUNDED
```

Comment:
Status should only be changed later when webhook is received as at this point only the refund request has been issued. (In practical terms, there might be considerable delay if the reservee needs to enter bank account number in Paytrail to actually receive the refund; otherwise the webhook call will be almost immediate.)

Reply:
Fixed.